### PR TITLE
Configurable ImageIO output format

### LINKED ISSUE #64: 32-bit instead of 8-bit PNG output
Rendering meta tiles with mod_tile/renderd always made my hillshade look ugly. This results in the 8 bit PNGs, which Mapnik doesn't rasterize nicely. The difference can be seen here:
8-bit: png256
32-bit: png

Only two lines have to be changed in src/gen_tile.c:
Substitute both "png256" with "png"

There are two possible ways:
- Introduce option in user config file (/usr/local/etc/renderd.conf) for bit depth
- Improve Mapnik. Imagemagick can produce nice looking 8-bit PNGs for example by "convert -type Palette -quality 90 [...]"

### PR DESCRIPTION
**Pull request by timautin**
Added the possibility to provide ImageIO's output format in the TYPE string of the renderd.conf file.
For instance:
TYPE=png image/png
TYPE=png image/png png256
TYPE=png image/png png24
TYPE=png image/png png

This should be backward compatible with existing renderd.conf files.

(closes #64)

### CODE REVIEW

File: src/gen_tile.cpp [hunk 1]
```diff
@@ -274,13 +275,14 @@ static enum protoCmd render(struct xmlmapconfig * map, int x, int y, int z, char
 #else
             mapnik::image_view<mapnik::image_data_32> vw(xx * map->tilesize, yy * map->tilesize, map->tilesize, map->tilesize, buf.data());
 #endif
-            tiles.set(xx, yy, save_to_string(vw, "png256"));
+            //tiles.set(xx, yy, save_to_string(vw, "png256"));
```
Comment (Firefishy) [inline src/gen_tile.cpp:278]: Likely worth removing the commented out line.

**Comment by xamanu** [comment]
Love this! Can you rebase this on latest `master` and add some documentation around this?

**Comment by timautin** [comment]
@xamanu Sure, are you going to merge the PR? I'm not used to rebase, would you be fine with a merge of master in my branch instead?
Would some doc in https://github.com/openstreetmap/mod_tile/blob/master/examples/config/renderd/renderd.conf.dist be fine?

**Comment by xamanu** [comment]
@timautin, no, please do a rebase not a merge, which would mix all things up. It is really just a:

* `git checkout master && git pull`
* `git checkout feature-configurableOutputFormat && git rebase master`
* `git push --force`

Yes, adding it to the `renderd.conf.dist` file would be a good fit.

**Comment by timautin** [comment]
Alright, that's pretty straightforward. But since I have deleted this project since I created the PR, I'm doing this:

git clone https://github.com/timautin/mod_tile.git && cd mod_tile
git remote add upstream https://github.com/openstreetmap/mod_tile.git
git fetch upstream
git checkout feature-configurableOutputFormat
git rebase upstream/master
[update the doc and remove the comment]
git push --force

Good for you?

**Comment by xamanu** [comment]
Looks good!

**Comment by xamanu** [comment]
* Unfortunately, it seems, the changes introduced here lead to failing tests. Please have a look at what is going on here.
* Today, I merged the changes for code formatting (#228), this will require another rebase from your side. This will cause conflicts because the style has changed, but these should be easy to fix, as no logic has changed.

Thank you for your contribution and for improving `mod_tile`.

**Comment by hummeltech** [comment]
Hello @xamanu & @timautin, I picked this up in my fork since it would also resolve #301 (in addition to the previous work here in resolving #64). I also added a very basic test to ensure WEBP images are being returned as expected. Let me know if this is something you'd be interested in.

https://github.com/hummeltech/mod_tile/compare/master...hummeltech:mod_tile:feature-configurableOutputFormat

**Comment by motiejus** [comment]
> Hello @xamanu & @timautin, I picked this up in my fork since it would also resolve #301 (in addition to the great work already here in resolving #64). I also added a very basic test to ensure `webp` & `png32` images are being returned as expected. Let me know if this is something you'd be interested in.
> 
> [hummeltech/mod_tile@openstreetmap:master...hummeltech:mod_tile:feature-configurableOutputFormat](https://github.com/hummeltech/mod_tile/compare/openstreetmap:master...hummeltech:mod_tile:feature-configurableOutputFormat)
> 
> Tests passing: https://github.com/hummeltech/mod_tile/actions/runs/4377116812

Thanks for your work on it, I was about to do the same. :)

Are you planning to open a PR for this? Your links above only link to the diffs.

**Comment by hummeltech** [comment]
Sure, I could open a new pull request for it, I didn't want to step on anyone's toes, but it seems like it's been over two years since there was any activity on this.

**Comment by motiejus** [comment]
> Sure, I suppose I could open a new pull request for it, I didn't want to step on anyone's toes, but it seems like it's been over two years since there was any activity on this.

Thanks, appreciated!

Once you open your PR, I will try it out with my use case and report the results.

**Comment by xamanu** [comment]
Went in with #318

**Pull request closed without merging**